### PR TITLE
Various improvements

### PR DESCRIPTION
**Pull request by top-sekret**
Various improvements enhancing psptoolchain

### CODE REVIEW

**Comment by dsieradzki** [comment]
Hello, you have to also fix "prepare scripts", I had to manually install isl on MacOS (brew install isl)

**Comment by top-sekret** [comment]
I think it tries to load both includes and libs from the build root (the use of ``$(pwd)`` is misleading, I'd introduce variables like ``$srcdir`` and ``$builddir``).

**Comment by top-sekret** [comment]
I also think that if the same fix were applied to GMP, MPFR and MPC, there would be no need to install it manually on the system.

**Comment by carstene1ns** [comment]
Have you tested the new gcc?

**Comment by top-sekret** [comment]
@carstene1ns yes, the SDK was built successfully, newlib was built successfully and so were the libraries

I mean, there was not a big load of changes in the patches. I've been trying to port new binutils, but this one is a bigger deal.

Also, I think GCC 8 is a good idea, because it can highlight visually (not only by character positions) which part of the line causes a warning/error.

File: patches/gcc-8.2.0-PSP.patch [hunk 1]
```diff
@@ -320,16 +320,16 @@ diff -Nru gcc-4.9.3/gcc/config/mips/mips.c gcc-4.9.3-psp/gcc/config/mips/mips.c
    /* Builtin functions for ST Microelectronics Loongson-2E/2F cores.  */
    LOONGSON_BUILTIN (packsswh, MIPS_V4HI_FTYPE_V2SI_V2SI),
    LOONGSON_BUILTIN (packsshb, MIPS_V8QI_FTYPE_V4HI_V4HI),
-@@ -14273,6 +14342,8 @@
- /* Standard mode-based argument types.  */
+@@ -16441,6 +16510,8 @@
  #define MIPS_ATYPE_UQI unsigned_intQI_type_node
+ #define MIPS_ATYPE_HI intHI_type_node
```
Comment (carstene1ns) [inline patches/gcc-8.2.0-PSP.patch]: This one has been added twice.

**Comment by carstene1ns** [comment]
Sure, using a new GCC brings a lot of benefits. Would be nice to know, if a compiled example works on hardware.

**Comment by top-sekret** [comment]
I would have to look for my hardware :laughing:. Anyway, I'm just finishing getting rid of some of errors you've highlighted.

**Comment by dsieradzki** [comment]
Unfortunately I have compilation error on MacOS High Sierra 10.13.4:
`xgcc: error: addsf3: No such file or directory
make[2]: *** [addsf3.o] Error 1
make[2]: *** Waiting for unfinished jobs....
xgcc: error: subsf3: No such file or directory
make[2]: *** [subsf3.o] Error 1
xgcc: error: negsf2: No such file or directory
make[2]: *** [negsf2.o] Error 1
 `

**Comment by top-sekret** [comment]
@dsieradzki that's worth some investigation. Try building a compiler manually (get the tarball, configure and make it manually) to see if it is an issue in macOS or in my build scripts.

**Comment by carstene1ns** [comment]
Might be a race condition/dependency issue with parallel builds.

**Comment by top-sekret** [comment]
@carstene1ns I'm building a huge script update and I opt for building one task at a time by default and only selecting more tasks if requested.

**Comment by top-sekret** [comment]
OK I've pushed the scripting system rework

**Comment by top-sekret** [comment]
I've created a Docker image for building PSP toolchain (**not** containing PSP toolchain). It can be used to build and run PSP toolchain in a deterministic, lean and supported environment.

You can view it [here](https://hub.docker.com/r/topsekretpl/psptoolchain-ci), but as I said these are only the dependencies of PSP toolchain, not the actual toolchain.

**Comment by joel16** [comment]
For the record I just want to state that I've been using top-sekret's modified toolchain for a while without any problems.

**Comment by PrimordialHelios** [comment]
This would be fantastic! A newer GCC would be absolutely lovely. It would also close #105 :P

**Comment by top-sekret** [comment]
If the newer GCC version fixed it, then of course

**Comment by top-sekret** [comment]
I think I'll dispose of insight, as it's no longer maintained. I suggest using [Nemiver](https://wiki.gnome.org/Apps/Nemiver), which is more modern, maintained, looks better and does need to be compiled for PSP (you can use the option ``--gdb-binary`` to choose ``psp-gdb``)

**Comment by top-sekret** [comment]
And I'll try to forward the binutils patches to a version as recent as possible.

**Comment by carstene1ns** [comment]
Updating binutils is a lot more work than gcc unfortunately, so good luck!

**Comment by top-sekret** [comment]
@carstene1ns not if you step version-by-version

**Comment by top-sekret** [comment]
They completely changed the parser in 2.24. I think someone with a deeper knowledge of MIPS could just rewrite this patch from scratch on the latest binutils.

**Comment by top-sekret** [comment]
But I think we could just merge this into master and I'd put newer binutils in a separate pull request/issue.

**Comment by BenMcLean** [comment]
I still get this error when I try to install this version:

> Failed installing: angelscript bzip2 openal

**Comment by top-sekret** [comment]
Could you send some more logs?

**Comment by BenMcLean** [comment]
Sure. I can run it again and upload the log.

**Comment by BenMcLean** [comment]
OK here is the log!
https://gist.github.com/BenMcLean/4f0c4e3cbc4caf1d014ad0f2bca37853

**Comment by top-sekret** [comment]
I've reviewed your log and here are my observations:

 * I think the problem lies within psplibraries build scripts
 * The psplibraries build script hides some logs, it needs a major rewrite
 * In case of ``angelscript`` and ``openal``, something with cmake is broken
 * In case of ``bzip2``, the broken logging of psplibraries build script prevents us from doing anything

**Comment by BenMcLean** [comment]
I'd really appreciate anyone helping get this fixed! It would be nice to see new homebrews for the PSP and PSVita still being made.

**Comment by BenMcLean** [comment]
Would the start of the solution be to fork psplibraries and fix the logging first?

**Comment by top-sekret** [comment]
I see the repo has become active again. If I cherry-pick all relevant commits, will this get merged? @dbeef

**Comment by top-sekret** [comment]
Because the problem I see is that new changes are made against the old scripts, which in turn causes the scripts from this pull request to have conflicts.

**Comment by dbeef** [comment]
@top-sekret I was thinking about this PR, as you already noticed there are multiple conflicting changes introduced.

Thus you definitely have to rebase your GCC script changes, as it has been upgraded to 9.3.0, and the newlib as of now is shipped from [pspdev/newlib](https://github.com/pspdev/newlib/).

For the `.gitlab-ci.yml` file, I don't see the point in including it without actually setting up the CI.
Github supports so called _Actions_ and I imagine that this, connected with [pspdev/docker](https://github.com/pspdev/pspdev-docker) will eventually run the CI, but this is only to be done in the future.

Here's how I see it - as there is a ton of changes in shell scripts, a development branch would be created on `pspdev/psptoolchain`, merging your changes there (granted you resolve conflicts), spending some time testing for regressions and then finally merging to master.

I am calling @carstene1ns and @ooPo to discuss this, as this is only an opinion of mine.

**Comment by top-sekret** [comment]
@dbeef OK, I've updated GCC and Newlib. I've also removed GitLab CI.

**Comment by top-sekret** [comment]
I've ported all changes to my fork.

**Comment by dbeef** [comment]
Ok, so now @ooPo please create `development` branch out of the `legacy` one.
Then target branch for this PR must be changed to pspdev/development, will probably merge without conflicts.

**Comment by dbeef** [comment]
Merged to development branch. Half way done, now we need to do some testing. 

I invite you to the PSP homebrew discord server, there's a channel called `psp-toolchain` where relevant discussions take place.

https://discord.gg/bePrj9W

**Comment by carstene1ns** [comment]
I am in for merging specific changes, but I generally dislike stuff like "here, take my script. it is better than yours".
For example the readme stuff cannot be applied like it is (mentions specific stuff for this fork).
The merging of the gcc stage scripts and xz extraction are simple enough to add right away, but I am not fond of the "improved" dependency and preparation script.

**Comment by sharkwouter** [comment]
To add to what @carstene1ns is saying about the dependency script, yesterday night I found that the current dependency installation script for Ubuntu can be run in an automatic way, while the one in this repo will need user input. The current script for installing MacOS dependencies is really old and probably doesn't even work anymore, though, while the one in the PR should work better for that.

**Comment by top-sekret** [comment]
@carstene1ns I am sorry. I did not intend to express such an attitude, but I believe you may be right, so my apologies for it.

When it comes to the Docker image, I believe the official Docker image didn't exist back then when I wrote it. You're absolutely right that it needs to be fixed. I can even do it if you point me to the correct Docker image.

When it comes to the dependency script, I separated it into sections and filtered it a little. If it introduces some unnecessary dependencies or does not check for some required, then I can fix it.

I rewrote the preparation script to unify it for all platforms. Regardless whether your doubts concern the script itself or the dependency lists contained inside, just let me address these issues. @sharkwouter Could you show me at what stage it requests user input? Would it be fixed by adding the `-y` option to `apt`?

After all, that's the point of `development` branch — it's not ready yet and I'm ready to work and collaborate on it and fix any issues that pop up.

**Comment by top-sekret** [comment]
Also, I just noticed, the `else` part in code below should never occur, as `curl` is bundled on macOS. So it will only be triggered if someone messed up the system or `$PATH`.

https://github.com/pspdev/psptoolchain/blob/699b0b17eae36a3986cc44e05a12314392d215e5/prepare.sh#L126
```shell
			if command -v wget >/dev/null 2>&1
			then
				wget --continue -O libusb-legacy.tar.gz "$LIBUSB_URL"
			elif command -v curl >/dev/null 2>&1
			then
				curl --continue-at - -o libusb-legacy.tar.gz "$LIBUSB_URL"
			else
				cat >&2 <<_EOF_
Error: no suitable downloader found.
Install one of these:
  * wget (preferred) <https://www.gnu.org/software/wget/>
  * curl <https://curl.haxx.se/>
_EOF_
				exit 77
			fi
```

**Comment by sharkwouter** [comment]
> I rewrote the preparation script to unify it for all platforms. Regardless whether your doubts concern the script itself or the dependency lists contained inside, just let me address these issues. @sharkwouter Could you show me at what stage it requests user input? Would it be fixed by adding the `-y` option to `apt`?

Yes, it asks you if you're sure when running apt install. I wasn't sure if this is intended, the original script does this too. The difference is that you can pass arguments to the original script, so you can just run ``sudo ./prepare-debian-and-ubuntu.sh -y``. Not sure if that's what the script is called, but you get the idea.

**Pull request merged**